qiskit/circuit/quantumcircuit.py

```diff
@@ -3698,7 +3698,14 @@ def copy(self, name: str | None = None) -> typing.Self:
         Returns:
           QuantumCircuit: a deepcopy of the current circuit, with the specified name
         """
-        cpy = self.copy_empty_like(name)
+        if not (name is None or isinstance(name, str)):
+            raise TypeError(
+                f"invalid name for a circuit: '{name}'. The name must be a string or 'None'."
+            )
+        cpy = _copy.copy(self)
+
+        _copy_metadata(self, cpy, "alike")
+
         cpy._data = self._data.copy()
         return cpy
```

Comment:
How come this needed to change? It makes me a bit worried that it will be easy to get out-of-sync with `copy_empty_like` if it doesn't call it (a problem I'm relatively sure we had in the past) - the additional cost of making a new empty `CircuitData` should (in theory) be dwarfed by the Python-space `copy.copy` (though the current behaviour of always cloning the interners would change that).

Reply:
Right now the change isn't as needed. If we go through with #13686, we might need to separate them because of how adding registers to `CircuitData` might work. I'm okay with leaving it as it was for this one and then modifying it if we go with the aforementioned approach.

Reply:
Addressed in dc555c8